sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +380,487 @@ def transform(self, X):
+            missing_values='NaN',
+            imputation_order='monotone',
+            n_imputations=100,
+            n_burn_in=10,
+            estimator=None,
+            n_nearest_features=None,
+            initial_strategy="mean",
+            min_value=None,
+            max_value=None,
+            verbose=False,
+            random_state=None):
+
+        self.missing_values = missing_values
+        self.imputation_order = imputation_order
+        self.n_imputations = n_imputations
+        self.n_burn_in = n_burn_in
+        self.estimator = estimator
+        self.n_nearest_features = n_nearest_features
+        self.initial_strategy = initial_strategy
+        self.min_value = min_value
+        self.max_value = max_value
+        self.verbose = verbose
+        self.random_state = random_state
+
+    def _impute_one_feature(self,
+                            X_filled,
+                            mask_missing_values,
+                            feat_idx,
+                            neighbor_feat_idx,
+                            estimator=None,
+                            min_std=1e-6):
+        """Imputes a single feature from the others provided.
+
+        This function predicts the missing values of one of the features using
+        the current estimates of all the other features. The `estimator` must
+        support `return_std=True` in its ``predict`` method for this function
+        to work.
+
+        Parameters
+        ----------
+        X_filled : array-like
+            Input data with the most recent imputations.
+
+        mask_missing_values : array-like
+            Input data's missing indicator matrix.
+
+        feat_idx : integer
+            Index of the feature currently being imputed.
+
+        neighbor_feat_idx : array-like
+            Indices of the features to be used in imputing `feat_idx`
+
+        estimator : object, default=BayesianRidgeRegression
+            The estimator to use at this step of the round-robin imputation.
+            As of now, it must support `return_std` in its ``predict`` method.
+            If None, it will be fit, otherwise not.
+
+        min_std : float, optional (default=1e-5)
+            The smallest allowable standard deviation for the posterior
+            sampling step.
+
+        Returns
+        -------
+        X_filled : array-like
+            Input data with `X_filled[missing_row_mask, feat_idx]`
+            updated.
+
+        estimator : estimator with sklearn API
+            The fitted estimator used to impute
+            `X_filled[missing_row_mask, feat_idx]`.
+        """
+
+        # if nothing is missing, just return the default
+        rng = check_random_state(self.random_state)
+        if mask_missing_values[:, feat_idx].sum() == 0:
+            return X_filled, estimator
+        missing_row_mask = mask_missing_values[:, feat_idx]
+
+        # if no estimator provided, instantiate a new one and fit
+        if estimator is None:
```

Comment:
Sorry, I am not sure I totally understand. The logic here is that if this is called with no `estimator` it means we are in the `fit` phase, so we have to make one. Otherwise we can use what was passed in. This sounds very similar to what you were saying in your downstream comment: `I would create a private function which would take estimator_triplet as argument. If None, you will compute the triplet otherwise you use it. Like this you will be able to factorize fit_transform and transform easily.`

Reply:
Oh I misread the statement with `if np.std(y_train)> 0`. Can we always clone `self.estimators_` instead of a DummyRegressor? Somehow it should not matter since those regressor should work even with a constant y (at least soon)

Reply:
I made this change. Hopefully it will work across the board.

Reply:
As noted below, it did not work =( I think this will have to wait until BayesianRidge is explicitly fixed to work with zero variance inputs.

Reply:
So it should be fixed in

Reply:
Oh awesome. Should I try to rebase?